autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/importing/takeout/webflow/ImportGoogleBookmarksWebFlowViewModel.kt

```diff
@@ -0,0 +1,297 @@
+        } catch (e: Exception) {
+            logcat(WARN) { "Bookmark zip download failed: ${e.message}" }
+            _commands.emit(Command.ExitFlowAsFailure(UserCannotImportReason.DownloadError))
+        }
+    }
+
+    fun onBookmarkZipAvailable(zipData: ByteArray, folderName: String) {
+        viewModelScope.launch(dispatchers.io()) {
+            try {
+                val extractionResult = bookmarkExtractor.extractBookmarksHtml(zipData)
+                onBookmarksExtracted(extractionResult, folderName)
+            } catch (e: Exception) {
+                logcat(WARN) { "Error processing bookmark zip: ${e.message}" }
+                _commands.emit(Command.ExitFlowAsFailure(UserCannotImportReason.ErrorParsingBookmarks))
+            }
+        }
+    }
+
+    private suspend fun onBookmarksExtracted(extractionResult: ExtractionResult, folderName: String) {
+        when (extractionResult) {
+            is ExtractionResult.Success -> {
+                val importResult = takeoutBookmarkImporter.importBookmarks(
+                    htmlContent = extractionResult.bookmarkHtmlContent,
+                    destination = ImportFolder.Folder(folderName),
+                )
+                onBookmarksImported(importResult, folderName)
+            }
+
+            is ExtractionResult.Error -> {
+                logcat(WARN) { "Error extracting bookmarks from zip" }
+                _commands.emit(Command.ExitFlowAsFailure(UserCannotImportReason.ErrorParsingBookmarks))
+            }
+        }
+    }
+
+    private suspend fun onBookmarksImported(importResult: ImportSavedSitesResult, folderName: String) {
+        when (importResult) {
+            is ImportSavedSitesResult.Success -> {
+                val importedCount = importResult.savedSites.size
+                logcat { "Successfully imported $importedCount bookmarks into '$folderName' folder" }
+                _commands.emit(Command.ExitFlowWithSuccess(importedCount))
+            }
+
+            is ImportSavedSitesResult.Error -> {
+                logcat(WARN) { "Error importing bookmarks: ${importResult.exception.message}" }
+                _commands.emit(Command.ExitFlowAsFailure(UserCannotImportReason.ErrorParsingBookmarks))
+            }
+        }
+    }
+
+    fun onCloseButtonPressed(url: String?) {
+        terminateFlowAsCancellation(url ?: "unknown")
+    }
+
+    fun onBackButtonPressed(
+        url: String? = null,
+        canGoBack: Boolean = false,
+    ) {
+        // if WebView can't go back, then we're at the first stage or something's gone wrong. Either way, time to cancel out of the screen.
+        if (!canGoBack) {
+            terminateFlowAsCancellation(url ?: "unknown")
+            return
+        }
+
+        _viewState.value = ViewState.NavigatingBack
+    }
+
+    private fun terminateFlowAsCancellation(url: String) {
+        viewModelScope.launch {
+            _viewState.value = ViewState.UserCancelledImportFlow(stage = "unknown")
```

Comment:
TODO: Work out staging once JS->native messaging is implemented (handled in another branch)